File: controllers/whatsapp_controller.py
from flask import jsonify
import os
import requests

def whatsapp_webhook(request):
    mode = request.args.get("hub.mode")
    token = request.args.get("hub.verify_token")
    challenge = request.args.get("hub.challenge")
    
    if mode and token:
        if mode == "subscribe" and token == 'ABCD': #Colocar o token do environment
            return challenge, 200
        else:
            return "", 403

def enviar_mensagem(request):
    data = request.get_json()
    
    if data.get("object"):
        key_wpp = data.get("key_wpp")
        bot_name = data.get("bot_name")
        entry = data.get("entry", [{}])[0]
        changes = entry.get("changes", [{}])[0]
        value = changes.get("value", {})
        phone_number_id = value.get("metadata", {}).get("phone_number_id")
        messages = value.get("messages", [{}])[0]
        from_number = messages.get("from")
        
        #send_whatsapp_message(from_number, response_message, phone_number_id, key_wpp)
        return jsonify({"status": "success"}), 200
    
    return jsonify({"error": "Not Found"}), 404

def send_whatsapp_message(phone_number, message, phone_number_id, key_wpp):
    url = f"{os.getenv('FACEBOOK_API_URL')}{phone_number_id}/messages?access_token={key_wpp}"
    json_api = {
        "messaging_product": "whatsapp",
        "to": phone_number,
        "text": {"body": message},
    }
    try:
        response = requests.post(url, json=json_api)
        response.raise_for_status()
    except requests.exceptions.RequestException as error:
        print(error)

def send_first_message(phone_number, key_wpp, template_wpp, bot_name, phone_number_id):
    url = f"{os.getenv('FACEBOOK_API_URL')}{phone_number_id}/messages?access_token={key_wpp}"
    message = {
        "messaging_product": "whatsapp",
        "to": phone_number,
        "type": "template",
        "template": {
            "name": template_wpp,
            "language": {"code": "pt_BR"},
            "components": [
                {
                    "type": "body",
                    "parameters": [
                        {
                            "type": "text",
                            "text": bot_name,
                        }
                    ],
                }
            ],
        },
    }
    try:
        response = requests.post(url, json=message)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as error:
        return {"error": 404, "message": str(error), "details": response.text if response else "Sem resposta do servidor"}

File: src/controllers/whatsappController.js
const axios = require('axios');
require('dotenv').config();

const { MessageHistory, User, Contact } = require('../models');

const { Queue } = require('bullmq');
const IORedis = require('ioredis');
const connection = new IORedis({ host: process.env.REDIS_HOST, port: process.env.REDIS_PORT, maxRetriesPerRequest: null });
const notificationQueue = new Queue('notifications', { connection });

exports.webhook = async (req, res) => {
    const data = req.body;
    if (data.object) {
        const entry = data.entry?.[0];
        const changes = entry?.changes?.[0];
        const value = changes?.value || {};
        const statuses = value?.statuses?.[0];
        const status = statuses.status;
        const messageId = statuses.id;
        if (status && messageId) {
            const messageHistory = await MessageHistory.findOne({
                where: { messageId: messageId},
                raw: false
            });
            if (messageHistory) {
                messageHistory.messageStatus = status;
                await messageHistory.save();
            }
        }
        return res.status(200).json({ status: 'success' });
    }

    return res.status(404).json({ error: 'Not Found' });
};

exports.configureWebhook = async (req, res) => {
    let mode = req.query["hub.mode"];
    let token = req.query["hub.verify_token"];
    let challenge = req.query["hub.challenge"];
    if (mode && token) {
        const user = await User.findOne({ where: { email: token } })
        if (mode === "subscribe" && user) {
            return res.status(200).send(challenge);
        } else {
            return res.sendStatus(403);
        }
    }
    return res.sendStatus(403);
  };

exports.sendFirstMessage = async (req, res) => {
    var response = await sendFirstMessage(req.user.referenceUser ?? req.user.id, req.body);
    if (response.error) {
        return res.status(400).json({ error: response });
    }
    return res.status(200).json(response);
}

// Função para enviar uma mensagem via WhatsApp
async function sendWhatsappMessage(phoneNumber, message, phoneNumberId, keyWpp) {
    const url = `${process.env.FACEBOOK_API_URL}${phoneNumberId}/messages?access_token=${keyWpp}`;

    const jsonApi = {
        messaging_product: 'whatsapp',
        to: phoneNumber,
        text: { body: message },
    };

    try {
        const response = await axios.post(url, jsonApi);
        return response.data;
    } catch (error) {
        console.error('Error sending message:', error);
    }
}

async function sendFirstMessage(userId, body) {
    const message = {
        userId: userId,
        template: body
    }
    
    const fullContact = {
        phone: body.phone_number,
    }
    const payload = {
        type: 'whatsapp',
        contact: fullContact,
        message: message,
    };

    try {
        await notificationQueue.add('sendNotification', payload);
        return { success: true, status: 'queued' };
    } catch (error) {
        return {
            error: 'queue_error',
            message: error.message,
        };
    }
}